Plastic potential

albert-de-montserrat opened the pull request:
Add the type for plastic potential:
```julia
struct PlasticPotential{F1,F2,F3,F4,F5,F6,F7} <: AbstractPlasticPotential{Float64}
    Q::F1 # Plastic potential
    ∂Q∂τxx::F2
    ∂Q∂τyy::F3
    ∂Q∂τzz::F4
    ∂Q∂τyz::F5
    ∂Q∂τxz::F6
    ∂Q∂τxy::F7

    function PlasticPotential(;
        Q::F1=second_invariant,
        ∂Q∂τxx::F2=(τij) -> 0.5 * τij[1] / second_invariant(τij),
        ∂Q∂τyy::F3=(τij) -> 0.5 * τij[2] / second_invariant(τij),
        ∂Q∂τzz::F4=nothing,
        ∂Q∂τyz::F5=nothing,
        ∂Q∂τxz::F6=nothing,
        ∂Q∂τxy::F7=(τij) -> τij[3] / second_invariant(τij),
    ) where {F1,F2,F3,F4,F5,F6,F7}
        return new{F1,F2,F3,F4,F5,F6,F7}(
           Q, ∂Q∂τxx, ∂Q∂τyy, ∂Q∂τzz, ∂Q∂τyz, ∂Q∂τxz, ∂Q∂τxy
        )
    end
end
```
I decided to split it from the plastic types, to avoid rewritting a bunch of code due to `dimensionalize()` and `nondimensionalize()` not dealing well with anonymous functions.

The are two options for the user. One is to provide a function for the plastic potential `Q` and let `ForwardDiff.jl` deal with the gradient; or let the user provide the analytical derivatives for  `∂Q∂τij`. Examples of usage:

```julia
# Test plastic potential derivatives
    ## 2D
    τij = (1.0, 2.0, 3.0)
    fxx(τij) = 0.5 * τij[1]/second_invariant(τij)
    fyy(τij) = 0.5 * τij[2]/second_invariant(τij)
    fxy(τij) = τij[3]/second_invariant(τij)
    solution2D = [fxx(τij), fyy(τij), fxy(τij)]

    p1 = PlasticPotential(; ∂Q∂τxx=fxx, ∂Q∂τyy=fyy, ∂Q∂τxy=fxy)
    p2 = PlasticPotential(; ∂Q∂τxx=nothing, ∂Q∂τyy=nothing, ∂Q∂τxy=nothing)

    # using StaticArrays
    τij_static = @SVector [1.0, 2.0, 3.0]
    out1 = ∂Q∂τ(p1, τij_static)
    out2 = ∂Q∂τ(p2, τij_static)
    out1 == out2 == solution2D

    # using tuples
    τij_tuple = (1.0, 2.0, 3.0)
    out3 = ∂Q∂τ(p1, τij)
    out4 = ∂Q∂τ(p2, τij)
    out3 == out4 == Tuple(solution2D)

    ## 3D
    τij = (1.0, 2.0, 3.0, 4.0, 5.0, 6.0)
    gxx(τij) = 0.5 * τij[1]/second_invariant(τij)
    gyy(τij) = 0.5 * τij[2]/second_invariant(τij)
    gzz(τij) = 0.5 * τij[3]/second_invariant(τij)
    gyz(τij) = τij[4]/second_invariant(τij)
    gxz(τij) = τij[5]/second_invariant(τij)
    gxy(τij) = τij[6]/second_invariant(τij)
    solution3D = [gxx(τij), gyy(τij), gzz(τij), gyz(τij), gxz(τij), gxy(τij)]

    p1 = PlasticPotential(; ∂Q∂τxx=gxx, ∂Q∂τyy=gyy, ∂Q∂τzz=gzz, ∂Q∂τyz=gyz, ∂Q∂τxz=gxz, ∂Q∂τxy=gxy)
    p2 = PlasticPotential(; ∂Q∂τxx=nothing, ∂Q∂τyy=nothing, ∂Q∂τzz=nothing, ∂Q∂τyz=nothing, ∂Q∂τxz=nothing, ∂Q∂τxy=nothing)

    # using StaticArrays
    τij_static = @SVector [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]
    out1 = ∂Q∂τ(p1, τij_static)
    out2 = ∂Q∂τ(p2, τij_static)
    out1 == out2 == solution3D

    # using tuples
    τij_tuple = (1.0, 2.0, 3.0, 4.0, 5.0, 6.0)
    out3 = ∂Q∂τ(p1, τij)
    out4 = ∂Q∂τ(p2, τij)
   out3 == out4 == Tuple(solution3D)
```

boriskaus commented:
thanks this is an important addition! 
A question from my side is that users should normally not really have to deal with plastic potentials; it's used internally for the local plasticity corrections, and perhaps you need it only at the end to visualise the amount of plastic strainrate (if multiplied with the plastic multiplier). 
Also for a given yield function, the plastic flow potential is known (e.g. Drucker Prager), and should be hardcoded in `GeoParams`. In some cases we will have analytical expressions of the derivatives of Q, but in other cases that might indeed be too complicated to derive and AD is indeed helpful.

I'm wondering whether this is giving users perhaps too much flexibility, and has the risk of them defining flow potentials that do not fit the yield function (that could mess up things badly).

albert-de-montserrat commented:
I reworked it following your suggestions. I removed the `struct` and moved everything to multiple dispatch based on the yield criteria. The usage doesn't change much, a 2D example:
```julia
    τij = (1.0, 2.0, 3.0)
    fxx(τij) = 0.5 * τij[1]/second_invariant(τij)
    fyy(τij) = 0.5 * τij[2]/second_invariant(τij)
    fxy(τij) = τij[3]/second_invariant(τij)
    solution2D = [fxx(τij), fyy(τij), fxy(τij)]
 
    # using StaticArrays
    τij_static = @SVector [1.0, 2.0, 3.0]
    out1 = ∂Q∂τ(p, τij_static)
    @test out1 == solution2D # true
    @test compute_plasticpotential(p, τij_static) == ∂Q∂τ(p, τij_static) # true

    # using tuples
    τij_tuple = (1.0, 2.0, 3.0)
    out2 = ∂Q∂τ(p, τij_tuple)
    @test out2 == Tuple(solution2D) # true
    @test compute_plasticpotential(p, τij_tuple) == ∂Q∂τ(p, τij_tuple) # true

    # using AD
    Q = second_invariant # where second_invariant is a function
    ad1 = ∂Q∂τ(Q, τij_static)
    @test out1 == solution2D # true
    @test compute_plasticpotential(p, τij_static) == ∂Q∂τ(p, τij_static) # true
    ad2 = ∂Q∂τ(Q, τij_tuple)
    @test out2 == Tuple(solution2D) # true
    @test compute_plasticpotential(p, τij_tuple) == ∂Q∂τ(p, τij_tuple) # true
```

boriskaus reviewed (approved):
thanks a lot! I suppose we will also need a routine that computes plastic strainrate invariant, given \lambda and dQd\tau, but this can be part of a follow-up PR
